### PR TITLE
on_stats request option

### LINKED ISSUE #1152: Cannot get connection time in middleware
In previous versions it was easy to get the (curl) connection time via `$event->getTransferInfo()`.
In guzzle6 I can't find a good way to do this. It seems that this information is not even stored anywhere.

I am aware that including this additional information would break the elegance and purity of the code so I can live without it. But maybe I am missing sth and it would be possible to pass this information so it's accessible in a handler?

I am using guzzle with a [Symfony2] (https://github.com/pinkeen/ApiDebugBundle) bundle which simplifies API debugging  in a few projects. Having the connection time really helps with identifying network problems and slow API endpoints/calls.

### PR DESCRIPTION
**Pull request by mtdowling**
This is  WIP commit that adds the `on_stats` request option.

Closes #1152

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Looks pretty nice. :)

**Comment by mtdowling** [comment]
Do you think the TransferStats class actually needs accessor methods, or should it just use public properties?

**Comment by GrahamCampbell** [comment]
Private properties with accessors is the best design wise, but public properties might save time. I use public properties more and more as I find it makes things easy to test and fast to write.

**Comment by stefanruijsenaars** [comment]
Just out of interest, what still needs to happen here?

**Comment by mtdowling** [comment]
@stefanruijsenaars mostly just tests and docs

**Comment by mtdowling** [comment]
This feature is now complete. How does it look from a functionality point of view,@GrahamCampbell / @Crell?

File: tests/TransferStatsTest.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?php
+namespace GuzzleHttp\Tests;
+
+use GuzzleHttp\TransferStats;
+use GuzzleHttp\Psr7;
+
+class TransferStatsTest extends \PHPUnit_Framework_TestCase
```
Comment (dawehner) [inline tests/TransferStatsTest.php]: Just curious: you don't use @coversDefaultClass and @covers right?

File: tests/TransferStatsTest.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public function testHasData()
```
Comment (dawehner) [inline tests/TransferStatsTest.php]: This test method name sounds wrong, what about testTransferStats instead?

File: tests/PrepareBodyMiddlewareTest.php [hunk 3]
```diff
@@ -126,7 +126,7 @@ function (RequestInterface $request) {
             new Request('PUT', 'http://www.google.com', ['Expect' => 'Foo'], $bd),
             ['expect' => true]
         );
-        $this->assertInstanceOf('GuzzleHttp\Promise\FulfilledPromise', $p);
+        $this->assertInstanceOf('GuzzleHttp\Promise\PromiseInterface', $p);
```
Comment (dawehner) [inline tests/PrepareBodyMiddlewareTest.php]: I'm curious, did you considered to use PromiseInterface::class instead?

**Comment by dawehner** [comment]
The general code looks pretty sane for me

File: tests/TransferStatsTest.php (hunk 2 above)
Comment (liveoutloud2day) [inline tests/TransferStatsTest.php]: In the other tests, classes are named SomethingTest, the functions are called testSomething, so I think this is OK. (Not really sure, but that is what the file tests/ClientTest.php has in its code).

File: tests/PrepareBodyMiddlewareTest.php (hunk 3 above)
Comment (mtdowling) [inline tests/PrepareBodyMiddlewareTest.php]: I forgot about that feature, but it would be nice to update the tests to use it at some point.

**Comment by lussoluca** [comment]
It's possibile to extract that information from a middleware?

**Comment by dawehner** [comment]
Can't your middleware manipulate the $options passed into the further down middleware/handler and by that register a callable?

**Comment by lussoluca** [comment]
Oh yes, I can do that! Thanks

File: tests/PrepareBodyMiddlewareTest.php (hunk 3 above)
Comment (dawehner) [inline tests/PrepareBodyMiddlewareTest.php]: Here is one: https://github.com/guzzle/guzzle/pull/1253

**Pull request merged**